subsys/net/lib/coap/coap_client.c

```diff
@@ -0,0 +1,669 @@
+		return ret;
+	}
+
+	ret = sendto(client->fd, client->request.data, client->request.offset, 0,
+		     &client->address, sizeof(client->address));
+	if (ret < 0) {
+		LOG_ERR("Error sending a CoAP ACK-message");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int send_reset(struct coap_client *client, const struct coap_packet *req,
+		      uint8_t response_code)
+{
+	int ret;
+	uint16_t id;
+	uint8_t token[COAP_TOKEN_MAX_LEN];
+	uint8_t tkl;
+
+	id = coap_header_get_id(req);
+	tkl = response_code ? coap_header_get_token(req, token) : 0;
+	ret = coap_packet_init(&client->request, client->send_buf, MAX_COAP_MSG_LEN, COAP_VERSION,
+			       COAP_TYPE_RESET, tkl, token, response_code, id);
+
+	if (ret < 0) {
+		LOG_ERR("Error creating CoAP reset message");
+		return ret;
+	}
+
+	ret = sendto(client->fd, client->request.data, client->request.offset, 0,
+		     &client->address, sizeof(client->address));
+	if (ret < 0) {
+		LOG_ERR("Error sending CoAP reset message");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int handle_response(struct coap_client *client, const struct coap_packet *response)
+{
+	int ret = 0;
+	int response_type;
+	int block_option;
+	int block_num;
+	bool blockwise_transfer = false;
+	bool last_block = false;
+
+	/* Handle different types, ACK might be separate or piggybacked
+	 * CON and NCON contains a separate response, CON needs an empty response
+	 * CON request results as ACK and possibly separate CON or NCON response
+	 * NCON request results only as a separate CON or NCON message as there is no ACK
+	 * With RESET, just drop gloves and call the callback.
+	 */
+	response_type = coap_header_get_type(response);
+
+	/* Reset and Ack need to match the message ID with request */
+	if ((response_type == COAP_TYPE_ACK || response_type == COAP_TYPE_RESET) &&
+	     coap_header_get_id(response) != client->pending.id)  {
+		LOG_ERR("Unexpected ACK or Reset");
+		return -EFAULT;
+	} else if (response_type == COAP_TYPE_RESET) {
+		coap_pending_clear(&client->pending);
+	}
+
+	/* CON, NON_CON and piggybacked ACK need to match the token with original request */
+	uint16_t payload_len;
+	uint8_t response_code = coap_header_get_code(response);
+	const uint8_t *payload = coap_packet_get_payload(response, &payload_len);
+
+	/* Separate response */
+	if (payload_len == 0 && response_type == COAP_TYPE_ACK &&
+	    response_code == COAP_CODE_EMPTY) {
+		/* Set a timeout value and clear pending */
+		coap_pending_clear(&client->pending);
+		return 1;
+	}
+
+	/* Check for tokens */
+	if (!token_compare(client, response)) {
+		LOG_ERR("Not matching tokens, respond with reset");
+		ret = send_reset(client, response, COAP_RESPONSE_CODE_NOT_FOUND);
+		return 1;
+	}
+
+	/* Send ack for CON */
+	if (response_type == COAP_TYPE_CON) {
+		/* If separate response, send empty ack. */
+		uint8_t con_response = COAP_CODE_EMPTY;
+
+		if (client->pending.timeout != 0) {
+			con_response = COAP_RESPONSE_CODE_OK;
+		}
```

Comment:
Not sure I understand the logic behind this if here.

Reply:
`client->pending.timeout` can be used in this case to identify if the response is a separate response, as the pending has been cleared and we are using a separate timeout value for the separate response.

Reply:
But isn't `COAP_TYPE_CON` response **always** a separate response? Otherwise it'd be piggybacked ACK. And the ACK for the separate response should be empty, `COAP_RESPONSE_CODE_OK` is a response code (and we're not replying for a request here, just acknowledging).

Reply:
Actually the server may respond to a `COAP_TYPE_NON` with `COAP_TYPE_CON`, and as the server must not respond with `COAP_TYPE_ACK` in that case, the `COAP_TYPE_CON` answer isn't a separate message.

Reply:
https://www.rfc-editor.org/rfc/rfc7252#page-12

Reply:
Yeah, that's actually true, https://www.rfc-editor.org/rfc/rfc7252#section-5.2.2 states that also CON responses to NON -requests are also thought as separate responses. I'll make the change here.

Reply:
Done